rfcs/rdfa-editor-stage-1.md

```diff
@@ -0,0 +1,587 @@
+comes with a cost. Plugin authors which implement their own node types are operating on the lowest possible level, and 
+are expected to read documentation and adhere to conventions, just as someone assembling their own computer from parts can be expected
+to handle components with care and follow instructions.
+
+I also consider allowing plugins to extend the available node-types a very powerful and expressive tool. Rather than concerning
+oneself with node-structures, one could implement a TodoItem nodetype, let's say, with natural methods such as `setStatus()` and `schedule()`,
+rather than interacting with things like html attributes. All we would need is to implement (de)serialization logic into and from the new type,
+and we'd be able to work with the node and write commands in a natural way. Requiring immutability at this point gives us state-tracking and
+atomic updates "for free" (aka without polluting the node's apis with unwieldy mutator objects).
+
+
+
+```typescript
+class InsertListItemCommand extends Command<{range: ModelRange}, void> {
+
+  canExecute({controller}, {range}): boolean {
+    return range.parentContext.containsNodeOfType("listNode");
+  }
+  
+  execute(config: CommandConfig, args: {range: ModelRange}): void {
+    const {controller} = config;
+    const {range} = args;
+    if(this.canExecute()) {
+      const listNode: ListNode = range.parentContext.findNodesOfType("listNode").next();
+      const listIndex = listNode.indexFromRange(range);
+      const newListNode = listNode.insertNewItem(listIndex);
+      const {mapping} = controller.mutator.replaceNode(listNode, newListNode);
+      controller.mutator.mapSelectionAndSet(mapping);
+    }
+  }
+}
+```
+
+
+#### Editor
+
+The Editor class is a container for all relevant state which is needed to create and operate the editor. 
+Of this state, the Document and the Selection are typically the most important and as of now the only ones
+which are defined. But we leave open the possibility for extra state to be added by plugins.
+  
+Besides managing the state, it also provides the primary api for modifying that state.
+   
+```typescript
+interface Editor {
+  commands: Command[];
+  document: ModelNode;
+  selection: ModelSelection;
+  eventBus: EventBus;
+}
+```
+
+#### Controller
+
+The controller class is a simple wrapper around the editor class. Every consumer of the editor will receive its own
+unique controller instance. This is done to track the origin of changes back to the callsite, and enable 
+plugins to recognize when a change-event is triggered by an action they performed, so they can avoid infinite loops.
+
+```typescript
+interface Controller {
+  name: string;
+  executeCommand<A, R>(commandName: string, args: A): R;
+  registerCommand<A, R>(command: new(): Command<A,R>): void;
+  registerWidget(spec: WidgetSpec): void;
+  registerNodeType(spec: NodeSpec): void;
+  mutator: Mutator;
+  selection: ModelSelection;
+  emitEvent(event: EditorEvent);
```

Comment:
no emitDebounced here?

Reply:
the idea was to set the high-level goals, not define all the interfaces precisely. interface definitions are minimal and can be extended.